Auto-update repo-policy-compliance service

yhaliaw opened the pull request:
* Modify the update-runner-binary event to also update the repo-policy-compliance service.
* Add `python` as alias for `python3`

yhaliaw committed:
Add resource config integration tests

yhaliaw committed:
Use operator-workflow integration test

yhaliaw committed:
Fix bugs

yhaliaw committed:
Fix incorrect type

yhaliaw committed:
Add --charm-file option for integration test

yhaliaw committed:
Refactor integration test fixtures

yhaliaw committed:
Split integration tests into smaller tests

yhaliaw committed:
Fix typo

yhaliaw committed:
Tmp disable --charm-file

yhaliaw committed:
Debug

yhaliaw committed:
Cleanup

yhaliaw committed:
Add checking of resource profile content

yhaliaw committed:
Rename function

yhaliaw committed:
Auto-update repo-policy-compliance service

yhaliaw committed:
Fix unit tests

yhaliaw committed:
Fix typo

yhaliaw committed:
Remove exception

yhaliaw committed:
Add logging for LXD network

yhaliaw committed:
Add alias for Python in runner

yhaliaw committed:
Fix tests

yhaliaw committed:
Fix typo

yhaliaw committed:
Add end-to-end test of python alias

yhaliaw committed:
Fix detection of version change

yhaliaw committed:
Merge branch 'main' into feat/auto-update-services

yhaliaw committed:
Add symlink of python3 as python

yhaliaw committed:
Refactor update dependencies

yhaliaw committed:
Add integration test for update of charm service

yhaliaw committed:
Fix integration test bug

yhaliaw committed:
Wait on status in integration tests

yhaliaw committed:
Fix test condition

yhaliaw committed:
Debug

yhaliaw committed:
Debug

yhaliaw committed:
Debug

yhaliaw committed:
Refactor integration test structure

yhaliaw committed:
Split the integration test workflow

merkata commented on src/charm.py:
```diff
@@ -427,10 +427,11 @@ def _check_and_update_dependencies(self) -> bool:
             return False
 
+        # TODO: Remove debugging.
+        logger.info("%s %s", runner_info.download_url, self._stored.runner_bin_url)
```
Would it be possible to leave debugging on and only trigger it conditionally, I recall bitbucket pipelines let's you define variables on runs, so you might pass debug=on and it defaults to debug=off for instance..

yhaliaw replied:
This is converted to debug-level logging.

yhaliaw committed:
Fix integration test

yhaliaw committed:
Fix integration tests

yhaliaw committed:
Update naming

yhaliaw committed:
Update debug logging

merkata commented on src/charm.py:
```diff
@@ -606,7 +651,15 @@ def _install_deps(self) -> None:
-        # Install repo-policy-compliance package
+        old_version = execute_command(
+            [
+                "/usr/bin/pip",
```
The check for pip in the GH actions invokes `python3 -m pip`, perhaps it can be invoked like this here as well?

yhaliaw replied:
Updated the code.

merkata commented on src/runner.py:
```diff
@@ -536,6 +536,9 @@ def _configure_runner(self) -> None:
             )
 
+        # Make python an alias of python3
+        self.instance.execute(["ln", "-s", "/usr/bin/python3", "/usr/bin/python"])
```
If `/usr/bin/python` exists this command will fail, is this expected?

yhaliaw replied:
This is run on start a runner LXD instance. Since the LXD instance is created from a ubuntu image, `/usr/bin/python` should be empty.

merkata commented on tests/integration/helpers.py:
```diff
@@ -14,8 +14,43 @@
+    Args:
+        source: The git source to install the package. If none the package is removed.
+    """
+    action = await unit.run("pip show repo-policy-compliance")
```
Similar to above comments, GH action invokes `python3 -m pip`, the charm invokes `pip` with a full path, here it relies on pip being in `$PATH` - perhaps there can be one way of invocation?

yhaliaw replied:
Switched to `python3 -m pip`

merkata reviewed (dismissed):
Looks good, left some minor nit picks.

yhaliaw committed:
Use
Usage:
  /home/yhaliaw/.pyenv/versions/3.10.12/bin/python3 -m pip <command> [options]

Commands:
  install                     Install packages.
  download                    Download packages.
  uninstall                   Uninstall packages.
  freeze                      Output installed packages in requirements format.
  inspect                     Inspect the python environment.
  list                        List installed packages.
  show                        Show information about installed packages.
  check                       Verify installed packages have compatible dependencies.
  config                      Manage local and global configuration.
  search                      Search PyPI for packages.
  cache                       Inspect and manage pip's wheel cache.
  index                       Inspect information available from package indexes.
  wheel                       Build wheels from your requirements.
  hash                        Compute hashes of package archives.
  completion                  A helper command used for command completion.
  debug                       Show information useful for debugging.
  help                        Show help for commands.

General Options:
  -h, --help                  Show help.
  --debug                     Let unhandled exceptions propagate outside the
                              main subroutine, instead of logging them to
                              stderr.
  --isolated                  Run pip in an isolated mode, ignoring
                              environment variables and user configuration.
  --require-virtualenv        Allow pip to only run in a virtual environment;
                              exit with an error otherwise.
  --python <python>           Run pip with the specified Python interpreter.
  -v, --verbose               Give more output. Option is additive, and can be
                              used up to 3 times.
  -V, --version               Show version and exit.
  -q, --quiet                 Give less output. Option is additive, and can be
                              used up to 3 times (corresponding to WARNING,
                              ERROR, and CRITICAL logging levels).
  --log <path>                Path to a verbose appending log.
  --no-input                  Disable prompting for input.
  --keyring-provider <keyring_provider>
                              Enable the credential lookup via the keyring
                              library if user input is allowed. Specify which
                              mechanism to use [disabled, import, subprocess].
                              (default: disabled)
  --proxy <proxy>             Specify a proxy in the form
                              scheme://[user:passwd@]proxy.server:port.
  --retries <retries>         Maximum number of retries each connection should
                              attempt (default 5 times).
  --timeout <sec>             Set the socket timeout (default 15 seconds).
  --exists-action <action>    Default action when a path already exists:
                              (s)witch, (i)gnore, (w)ipe, (b)ackup, (a)bort.
  --trusted-host <hostname>   Mark this host or host:port pair as trusted,
                              even though it does not have valid or any HTTPS.
  --cert <path>               Path to PEM-encoded CA certificate bundle. If
                              provided, overrides the default. See 'SSL
                              Certificate Verification' in pip documentation
                              for more information.
  --client-cert <path>        Path to SSL client certificate, a single file
                              containing the private key and the certificate
                              in PEM format.
  --cache-dir <dir>           Store the cache data in <dir>.
  --no-cache-dir              Disable the cache.
  --disable-pip-version-check
                              Don't periodically check PyPI to determine
                              whether a new version of pip is available for
                              download. Implied with --no-index.
  --no-color                  Suppress colored output.
  --no-python-version-warning
                              Silence deprecation warnings for upcoming
                              unsupported Pythons.
  --use-feature <feature>     Enable new functionality, that may be backward
                              incompatible.
  --use-deprecated <feature>  Enable deprecated functionality, that will be
                              removed in the future.

yhaliaw committed:
Use
Usage:
  /home/yhaliaw/.pyenv/versions/3.10.12/bin/python3 -m pip <command> [options]

Commands:
  install                     Install packages.
  download                    Download packages.
  uninstall                   Uninstall packages.
  freeze                      Output installed packages in requirements format.
  inspect                     Inspect the python environment.
  list                        List installed packages.
  show                        Show information about installed packages.
  check                       Verify installed packages have compatible dependencies.
  config                      Manage local and global configuration.
  search                      Search PyPI for packages.
  cache                       Inspect and manage pip's wheel cache.
  index                       Inspect information available from package indexes.
  wheel                       Build wheels from your requirements.
  hash                        Compute hashes of package archives.
  completion                  A helper command used for command completion.
  debug                       Show information useful for debugging.
  help                        Show help for commands.

General Options:
  -h, --help                  Show help.
  --debug                     Let unhandled exceptions propagate outside the
                              main subroutine, instead of logging them to
                              stderr.
  --isolated                  Run pip in an isolated mode, ignoring
                              environment variables and user configuration.
  --require-virtualenv        Allow pip to only run in a virtual environment;
                              exit with an error otherwise.
  --python <python>           Run pip with the specified Python interpreter.
  -v, --verbose               Give more output. Option is additive, and can be
                              used up to 3 times.
  -V, --version               Show version and exit.
  -q, --quiet                 Give less output. Option is additive, and can be
                              used up to 3 times (corresponding to WARNING,
                              ERROR, and CRITICAL logging levels).
  --log <path>                Path to a verbose appending log.
  --no-input                  Disable prompting for input.
  --keyring-provider <keyring_provider>
                              Enable the credential lookup via the keyring
                              library if user input is allowed. Specify which
                              mechanism to use [disabled, import, subprocess].
                              (default: disabled)
  --proxy <proxy>             Specify a proxy in the form
                              scheme://[user:passwd@]proxy.server:port.
  --retries <retries>         Maximum number of retries each connection should
                              attempt (default 5 times).
  --timeout <sec>             Set the socket timeout (default 15 seconds).
  --exists-action <action>    Default action when a path already exists:
                              (s)witch, (i)gnore, (w)ipe, (b)ackup, (a)bort.
  --trusted-host <hostname>   Mark this host or host:port pair as trusted,
                              even though it does not have valid or any HTTPS.
  --cert <path>               Path to PEM-encoded CA certificate bundle. If
                              provided, overrides the default. See 'SSL
                              Certificate Verification' in pip documentation
                              for more information.
  --client-cert <path>        Path to SSL client certificate, a single file
                              containing the private key and the certificate
                              in PEM format.
  --cache-dir <dir>           Store the cache data in <dir>.
  --no-cache-dir              Disable the cache.
  --disable-pip-version-check
                              Don't periodically check PyPI to determine
                              whether a new version of pip is available for
                              download. Implied with --no-index.
  --no-color                  Suppress colored output.
  --no-python-version-warning
                              Silence deprecation warnings for upcoming
                              unsupported Pythons.
  --use-feature <feature>     Enable new functionality, that may be backward
                              incompatible.
  --use-deprecated <feature>  Enable deprecated functionality, that will be
                              removed in the future. in tests

yhaliaw committed:
Safe guard agianst existing /usr/bin/python

yhaliaw committed:
Fix missing comma

jdkandersson commented on src/runner.py:
```diff
@@ -536,6 +536,10 @@ def _configure_runner(self) -> None:
             )
 
+        # Make python an alias of python3
+        self.instance.execute(["rm", "/usr/bin/python3"])
```
Should this `rm` `/usr/bin/python` rather than `/usr/bin/python3`?

yhaliaw replied:
Thanks this is fixed.

yhaliaw committed:
Fix typo
